kimchi/src/alphas.rs

```diff
@@ -0,0 +1,276 @@
+//! This module implements an abstraction to keep track of the powers of alphas.
+//! As a recap, alpha is a challenge sent by the verifier in PLONK,
+//! and is used to segregate different constraints.
+//! It is important that different constraints use different powers of alpha,
+//! as otherwise they can interact and potentially cancel one another.
+//! (The proof is in the use of the Schwartz-Zipple lemma.)
+//! As such, we want two properties from this:
+//!
+//! - we should keep track of a mapping between type of constraint and range of powers
+//! - when powers of alphas are used, we should ensure that no more no less are used
+//!
+//! We use powers of alpha in two different places in the codebase:
+//!
+//! - when creating the index, we do not know alpha at this point so we
+//!   simply keep track of what constraints will use what powers
+//! - when creating a proof or verifying a proof, at this point we know alpha
+//!   so we can use the mapping we created during the creation of the index.
+//!
+//! For this to work, we use two types:
+//!
+//! - [Builder], which allows us to map constraints to powers
+//! - [Alphas], which you can derive from [Builder] and an `alpha`
+//!
+//! Both construction will enforce that you use all the powers of
+//! alphas that you register for constraint. This allows us to
+//! make sure that we compute the correct amounts, without reusing
+//! powers of alphas between constraints.
+
+use ark_ff::Field;
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::HashMap,
+    iter::{Cloned, Take},
+    ops::Range,
+    slice::Iter,
+    thread,
+};
+
+// ------------------------------------------
+
+/// A constraint type represents a polynomial that will be part of the final equation f (the circuit equation)
+#[derive(PartialEq, Eq, Clone, Hash, Debug, Serialize, Deserialize)]
+pub enum ConstraintType {
+    /// gates in the PLONK constraint system.
+    /// As gates are mutually exclusive (only a single selector polynomial
+    /// will be non-zero for a given point),
```

Comment:
<img width="1160" alt="Screen Shot 2022-02-07 at 2 53 37 PM" src="https://user-images.githubusercontent.com/1316043/152885869-77e31983-3ad9-4ccb-aa06-c76f336fc444.png">

I refer to the gate selectors, the second table from the left. You see that each

Reply:
<img width="1156" alt="Screen Shot 2022-02-07 at 3 06 48 PM" src="https://user-images.githubusercontent.com/1316043/152887270-8a43a616-c246-46ff-867b-85eba3d1ddf7.png">

sigmas should only have 7 columns :D

Reply:
aha cool because the sigmas refer to the IO registers I suppose
Ok, this diagram clarifies many things, but also compliments what I said above I think. Particularly, it is good to clarify that these things are known to the verifier (as you say, it is the circuit specification / relation to be proven). 
Now for the column sum you refer to for the gates, I think it is the opposite: I would say it is row-wise what has an all-1 sum. But anyway, I know what you mean. And yeah, I guess these selector polynomials are computed using lagrange, what will give them this linearly independence property. Together with the fact that the verifier has access to these, it all makes sense to me, thanks 👍